packages/smart-contracts/src/pn-ETH-to-DAI-proxy.sol

```diff
@@ -0,0 +1,166 @@
+     * condition is to first reduce the spender's allowance to 0 and set the
+     * desired value afterwards:
+     * https://github.com/ethereum/EIPs/issues/20#issuecomment-263524729
+     *
+     * Emits an {Approval} event.
+     */
+    function approve(address spender, uint256 amount) external returns (bool);
+
+    /**
+     * @dev Moves `amount` tokens from `sender` to `recipient` using the
+     * allowance mechanism. `amount` is then deducted from the caller's
+     * allowance.
+     *
+     * Returns a boolean value indicating whether the operation succeeded.
+     *
+     * Emits a {Transfer} event.
+     */
+    function transferFrom(address sender, address recipient, uint256 amount) external returns (bool);
+
+    /**
+     * @dev Emitted when `value` tokens are moved from one account (`from`) to
+     * another (`to`).
+     *
+     * Note that `value` may be zero.
+     */
+    event Transfer(address indexed from, address indexed to, uint256 value);
+
+    /**
+     * @dev Emitted when the allowance of a `spender` for an `owner` is set by
+     * a call to {approve}. `value` is the new allowance.
+     */
+    event Approval(address indexed owner, address indexed spender, uint256 value);
+}
+
+
+/**
+ * @title ERC20Proxy
+ * @notice This contract performs an ERC20 token transfer and stores a reference
+  */
+contract ERC20ProxyFromETH {
+  // Event to declare a transfer with a reference
+  event TransferWithReference(
+    address tokenAddress,
+    address to,
+    uint256 amount,
+    bytes indexed paymentReference
+  );
+
+
+  address internal constant UNISWAP_ROUTER_ADDRESS = 0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D ;
+
+  IUniswapV2Router02 public uniswapRouter;
+  address private multiDaiKovan = 0x4F96Fe3b7A6Cf9725f59d353F723c1bDb64CA6Aa;
+
+  constructor() public {
+    uniswapRouter = IUniswapV2Router02(UNISWAP_ROUTER_ADDRESS);
+  }
+
+  function convertEthToDai(uint daiAmount) public payable {
+    uint deadline = now + 15; // using 'now' for convenience, for mainnet pass deadline from frontend!
+    uniswapRouter.swapETHForExactTokens.value(msg.value)(daiAmount, getPathForETHtoDAI(), address(this), deadline);
+    
+    // refund leftover ETH to user
+    msg.sender.call.value(address(this).balance)("");
+  }
+  
+  
+  function getPathForETHtoDAI() private view returns (address[] memory) {
+    address[] memory path = new address[](2);
+    path[0] = uniswapRouter.WETH();
+    path[1] = multiDaiKovan;
+    
+    return path;
+  }
+  
+  /**
+  * @notice Performs a DAI token transfer from ETH with a reference
+  * @param _to Transfer recipient
+  * @param _amount Amount to transfer
+  * @param _paymentReference Reference of the payment related
+  */
+  function transferDAIFromETHWithReference(
+    //address _tokenAddress,
+    address _to,
+    uint256 _amount,
+    bytes calldata _paymentReference
+  )
+    external 
+    payable
+  {
+    IERC20 erc20 = IERC20(multiDaiKovan);
+    convertEthToDai(_amount);
```

Comment:
shouldn't this be checked with `require` ?

Reply:
Yes, maybe. What's weird is that the Uniswap contract will reject if you pay... **too much** ETH. I have not tried yet to pay not enough. Many edge cases to consider!

Reply:
Now I understand better how everything works so I should be able to test the necessity of require and be sure.